compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ForLoopsLowering.kt

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.common.lower.loops
+
+import org.jetbrains.kotlin.backend.common.CommonBackendContext
+import org.jetbrains.kotlin.backend.common.FileLoweringPass
+import org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.lower.irIfThen
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.ir.IrStatement
+import org.jetbrains.kotlin.ir.builders.irCallOp
+import org.jetbrains.kotlin.ir.builders.irGet
+import org.jetbrains.kotlin.ir.builders.irSetVar
+import org.jetbrains.kotlin.ir.declarations.IrFile
+import org.jetbrains.kotlin.ir.declarations.IrVariable
+import org.jetbrains.kotlin.ir.expressions.*
+import org.jetbrains.kotlin.ir.expressions.impl.IrCompositeImpl
+import org.jetbrains.kotlin.ir.symbols.IrVariableSymbol
+import org.jetbrains.kotlin.ir.types.getClass
+import org.jetbrains.kotlin.ir.types.toKotlinType
+import org.jetbrains.kotlin.ir.util.functions
+import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
+import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
+
+val forLoopsPhase = makeIrFilePhase(
+    ::ForLoopsLowering,
+    name = "ForLoopsLowering",
+    description = "For loops lowering"
+)
+
+/**
+ * This lowering pass optimizes for-loops.
+ *
+ * Replace iteration over progressions (e.g., X.indices, a..b) and arrays with
+ * a simple while loop over primitive induction variable.
+ *
+ * For example, this loop:
+ *
+ * ```
+ * for (i in first..last step foo) { ... }
+ * ```
+ *
+ * is represented in IR in such a manner:
+ *
+ * ```
+ * val it = (first..last step foo).iterator()
```

Comment:
I would prefer to supports several different templates here (and seems they could vary on backends) with less tempory variables:
1. For totally safe cases with no overflow and non-empty interval, e.g. fir (i in 1..20), generate:
```
var i = first
while (i <= first) {
...
 i++
}
```
2. for other cases (this template looks like current jvm-backend work)
```
var i = first
  if (i <= last) { 
      while(true) {
 
          ...
 
           overflow check if required (closed bound) and break
           i++
      }
  }
```

Reply:
@punzki It'm suggesting very conservative approach (maybe we need some investigations with benchmarking) cause I don't know how good different JVM will optimize your template

Reply:
@dzharkov @udalov @dnpetrov Any thoughts?

Reply:
I agree, I would also prefer even more optimized version of the loop in case the right bound is a constant expression, similarly to how the current JVM backend manages to optimize it. But I'm not opposed to the current @punzki 's implementation either; it's OK to postpone this since it's not the _most_ practical use case of for loops. Maybe we need to ensure that there are bytecode text tests on this optimization though (that check that there's only one `if_icmp..` for a constant-int for loop)

Reply:
> I would prefer to supports several different templates here (and seems they could vary on backends) with less tempory variables:

@max-kammerer I noted this in a [TODO in `HeaderProcessor.kt`](https://github.com/JetBrains/kotlin/pull/2164/files#diff-8070e1cd08057790e0480b78ebc62042R78):
```
// TODO: Build while loop (instead of do-while) where possible, e.g., in "until" ranges, or when "last" is known to be < MAX_VALUE
```

I chose to leave it as a TODO since the original `ForLoopsLowering` didn't specifically handle the "no possible overflow" case. I do have a change in-progress to handle this. Would you prefer I include this change in this PR? Or in a future PR? I would personally prefer to push this to a different PR (makes reviews and my life simpler). :smile:

> Maybe we need to ensure that there are bytecode text tests on this optimization though (that check that there's only one if_icmp.. for a constant-int for loop)

@udalov There are in fact assertions to that effect in `forInRangeTo(Char|Long)Const.kt` and the tests are still failing (and disabled) for that very reason.

Reply:
@punzki 
Just note: in old backend we are awraing overflow only on increment operation, 
with always close bounds additional operation is introduced `last = rightOpenBound -1` and it would also can cause overflow, e.g. `for (i in 1..MIN_INT)`, so always closed bounds could add some complexity in generated bytecode.

Further optimizations could be added as separate PR. But we need to clarify situation about number of overflow checks and reduce its count.

Reply:
@max-kammerer Right, that's already considered in this PR. For `until` loops (see `UntilHandler`), there is an additional condition in the "emptiness check": `inductionVariable <= last && bound > MIN_VALUE`.

Reply:
@punzki Yes, but I'm awaring about nested loops:
```
for (ext in smth) {
   for (i in 1 until ext) {
  }
}
```
In such case we would check additional condition on every external loop iteration. But this can be avoided if we would keep information about interval bound - closed/opened.

Reply:
@max-kammerer I'm having trouble getting your point. Maybe you can explain in the form of the lowered for-loop?

The original `ForLoopsLowering` implementation distinguished between closed/open bounds and [always added the additional `bound > MIN_VALUE` condition](https://github.com/JetBrains/kotlin/pull/2164/commits/a1e4ab6a46d8f486a12f53cba6c17bcf859e55a7#diff-f2ebc3530ff622def330bd11ef340f7eL208). In your example, since we don't know what the value of `ext` could be, then we have to always add the extra condition. I don't see how it could be avoided by keeping the closed/open information.

Reply:
Alternatively, instead of computing `last = rightOpenBound - 1`, we can change the comparison operators and structure of the loop depending on the closed/open information. Is this what you were describing? Something like:

```
// 1. If right bound is OPEN (i.e., "until" loop, array, ".indices" loop), we can use a "while" loop:
var inductionVar = leftBound
while (inductionVar < rightBound) {
  val loopVar = inductionVar
  inductionVar += step
  // Loop body
}

// 2. If induction variable can NOT overflow, and right bound is CLOSED (I can include this in this PR since we are doing "while" loops above anyway)
var inductionVar = leftBound
while (inductionVar <= rightBound) {
  val loopVar = inductionVar
  inductionVar += step
  // Loop body
}

// 3. Default loop structure:
var inductionVar = leftBound
if (inductionVar <= rightBound) {
  do {
    val loopVar = inductionVar
    inductionVar += step
    // Loop body
  } while (inductionVar != rightBound)
}
```

Reply:
@max-kammerer I have a change that keeps the open/closed bound info and modifying the loop condition operator based on that (`<=` for closed, `<` for open). With this, the additional `bound > MIN_VALUE` condition is no longer necessary. If you think this is the way to go, let me know and I'll add it to the PR.

Personally, I think this is the right approach and it is much simpler. I should have done this in the first place but I got anchored on how the original `ForLoopsLowering` handled `until` loops. The old JVM backend just changes the operator based on open/closed.

Reply:
@punzki Yes, you catch my thoughts! (I meant additional condition about 'bound > MIN_VALUE'.)
And sorry for incomprehensibility from my side (last days I'm a bit sickly).

Reply:
@max-kammerer No trouble at all, hope you are better now! Updated the PR with changes to use while loops for possible, and use appropriate operator based on open/closed.